sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,788 @@
+):
+    # The Cython routines used require contiguous arrays
+    X = np.asarray(X, order="C")
+
+    # Get distance to kth nearest neighbour
+    nbrs = NearestNeighbors(
+        n_neighbors=min_samples,
+        algorithm=algo,
+        leaf_size=leaf_size,
+        metric=metric,
+        metric_params=metric_params,
+        n_jobs=n_jobs,
+        p=None,
+    ).fit(X)
+
+    neighbors_distances, _ = nbrs.kneighbors(X, min_samples, return_distance=True)
+    core_distances = np.ascontiguousarray(neighbors_distances[:, -1])
+    dist_metric = DistanceMetric.get_metric(metric, **metric_params)
+
+    # Mutual reachability distance is implicit in mst_from_data_matrix
+    min_spanning_tree = mst_from_data_matrix(X, core_distances, dist_metric, alpha)
+    return _process_mst(min_spanning_tree)
+
+
+def remap_single_linkage_tree(tree, internal_to_raw, non_finite):
+    """
+    Takes an internal single_linkage_tree structure and adds back in a set of points
+    that were initially detected as non-finite and returns that new tree.
+    These points will all be merged into the final node at np.inf distance and
+    considered noise points.
+
+    Parameters
+    ----------
+    tree: single_linkage_tree
+    internal_to_raw: dict
+        a mapping from internal integer index to the raw integer index
+    finite_index: ndarray
+        Boolean array of which entries in the raw data were finite
+    """
+    finite_count = len(internal_to_raw)
+
+    outlier_count = len(non_finite)
+    for i, _ in enumerate(tree):
+        left = tree[i]["left_node"]
+        right = tree[i]["right_node"]
+
+        if left < finite_count:
+            tree[i]["left_node"] = internal_to_raw[left]
+        else:
+            tree[i]["left_node"] = left + outlier_count
+        if right < finite_count:
+            tree[i]["right_node"] = internal_to_raw[right]
+        else:
+            tree[i]["right_node"] = right + outlier_count
+
+    outlier_tree = np.zeros(len(non_finite), dtype=HIERARCHY_dtype)
+    last_cluster_id = max(
+        tree[tree.shape[0] - 1]["left_node"], tree[tree.shape[0] - 1]["right_node"]
+    )
+    last_cluster_size = tree[tree.shape[0] - 1]["cluster_size"]
+    for i, outlier in enumerate(non_finite):
+        outlier_tree[i] = (outlier, last_cluster_id + 1, np.inf, last_cluster_size + 1)
+        last_cluster_id += 1
+        last_cluster_size += 1
+    tree = np.concatenate([tree, outlier_tree])
+    return tree
+
+
+def _get_finite_row_indices(matrix):
+    """
+    Returns the indices of the purely finite rows of a
+    sparse matrix or dense ndarray
+    """
+    if issparse(matrix):
+        row_indices = np.array(
+            [i for i, row in enumerate(matrix.tolil().data) if np.all(np.isfinite(row))]
+        )
+    else:
+        (row_indices,) = np.isfinite(matrix.sum(axis=1)).nonzero()
+    return row_indices
+
+
+class HDBSCAN(ClusterMixin, BaseEstimator):
+    """Cluster data using hierarchical density-based clustering.
```

Comment:
would also be nice to have a link to the example where DBSCAN and HDBSCAN are compared here, instead of just relying on the auto detection of all examples, which doesn't make examples really discoverable.

Reply:
Done!